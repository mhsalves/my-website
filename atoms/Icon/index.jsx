import React from 'react';
import {
  bool, number, oneOf, string,
} from 'prop-types';

import IconSvg from './styles';

function Icon(props) {
  const {
    glyph, size, ...other
  } = props;

  return (
    <IconSvg {...other} size={size}>
      <use href={`/static/svg/${size}.svg#${glyph}`} />
    </IconSvg>
  );
}

Icon.propTypes = {
  glyph: string.isRequired,
  size: oneOf(['small', 'base', 'large']),
  color: string,
  beforeText: bool,
  rotate: number,
  spin: bool,
};

Icon.defaultProps = {
  size: 'base',
  color: '',
  beforeText: false,
  rotate: null,
  spin: false,
};

export default Icon;
