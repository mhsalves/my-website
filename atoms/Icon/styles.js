import styled, { css, keyframes } from 'styled-components';
import colorUtil from '../../helpers/colorUtils';

const sizeSmall = css`
  width: ${({ theme: { space } }) => space.small};
  height: ${({ theme: { space } }) => space.small};
`;

const sizeBase = css`
  width: ${({ theme: { space } }) => space.base};
  height: ${({ theme: { space } }) => space.base};
`;

const sizeLarge = css`
  width: ${({ theme: { space } }) => space.large};
  height: ${({ theme: { space } }) => space.large};
`;

const beforeTextSpace = css`
  margin-right: ${({ theme: { space } }) => space.micro};
`;

const spinAnimation = keyframes`
  from { transform: rotate(0deg); }
  to { transform: rotate(-360deg); }
`;

const IconSvg = styled.svg`
  ${({ size }) => size === 'small' && sizeSmall}
  ${({ size }) => size === 'base' && sizeBase}
  ${({ size }) => size === 'large' && sizeLarge}

  ${({ beforeText }) => beforeText && beforeTextSpace};

  ${({ color, theme }) => color && css`color: ${colorUtil(color, theme)};`}
  ${({ rotate }) => rotate && css`transform: rotate(${rotate}deg);`}
  ${({ spin }) => spin && css`animation: ${spinAnimation} 1.8s linear infinite;`}
`;

export default IconSvg;
