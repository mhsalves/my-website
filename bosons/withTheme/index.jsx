import React from 'react';
import { ThemeProvider } from 'styled-components';
import baseTheme from '../themes/base';

const withTheme = Component => props => (
  <ThemeProvider theme={baseTheme}>
    <Component {...props} />
  </ThemeProvider>
);

export default withTheme;
