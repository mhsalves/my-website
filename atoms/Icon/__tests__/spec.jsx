import React from 'react';
import renderer from 'react-test-renderer';

import Icon from '../index';

import withTheme from '../../../bosons/withTheme';

const ComponentWithTheme = withTheme(Icon);
const glyph = 'glyph';

describe('given: Icon', () => {
  test('when: has no size', () => {
    const snap = renderer.create(<ComponentWithTheme glyph={glyph} size="base" />).toJSON();
    const defaultSnap = renderer.create(<ComponentWithTheme glyph={glyph} />).toJSON();
    expect(snap).toEqual(defaultSnap);
  });

  test('when: is small', () => {
    const component = renderer.create(
      <ComponentWithTheme glyph={glyph} size="small" />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('when: is large', () => {
    const component = renderer.create(
      <ComponentWithTheme glyph={glyph} size="large" />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('when: has color', () => {
    const component = renderer.create(
      <ComponentWithTheme glyph={glyph} color="primary" />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('when: has color and variant', () => {
    const component = renderer.create(
      <ComponentWithTheme glyph={glyph} color="coal default" />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('when: is before a text', () => {
    const component = renderer.create(
      <ComponentWithTheme glyph={glyph} beforeText />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('when: has rotation', () => {
    const component = renderer.create(
      <ComponentWithTheme glyph={glyph} rotate={90} />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  test('when: is spinning', () => {
    const component = renderer.create(
      <ComponentWithTheme glyph={glyph} spin />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });
});
