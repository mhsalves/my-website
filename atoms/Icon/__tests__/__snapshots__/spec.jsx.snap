// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given: Icon when: has color 1`] = `
.c0 {
  width: 24px;
  height: 24px;
  color: #BD5D38;
}

<svg
  className="c0"
  color="primary"
  rotate={null}
  size="base"
>
  <use
    href="/static/svg/base.svg#glyph"
  />
</svg>
`;

exports[`given: Icon when: has color and variant 1`] = `
.c0 {
  width: 24px;
  height: 24px;
  color: #868E96;
}

<svg
  className="c0"
  color="coal default"
  rotate={null}
  size="base"
>
  <use
    href="/static/svg/base.svg#glyph"
  />
</svg>
`;

exports[`given: Icon when: has rotation 1`] = `
.c0 {
  width: 24px;
  height: 24px;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

<svg
  className="c0"
  color=""
  rotate={90}
  size="base"
>
  <use
    href="/static/svg/base.svg#glyph"
  />
</svg>
`;

exports[`given: Icon when: is before a text 1`] = `
.c0 {
  width: 24px;
  height: 24px;
  margin-right: 8px;
}

<svg
  className="c0"
  color=""
  rotate={null}
  size="base"
>
  <use
    href="/static/svg/base.svg#glyph"
  />
</svg>
`;

exports[`given: Icon when: is large 1`] = `
.c0 {
  width: 32px;
  height: 32px;
}

<svg
  className="c0"
  color=""
  rotate={null}
  size="large"
>
  <use
    href="/static/svg/large.svg#glyph"
  />
</svg>
`;

exports[`given: Icon when: is small 1`] = `
.c0 {
  width: 16px;
  height: 16px;
}

<svg
  className="c0"
  color=""
  rotate={null}
  size="small"
>
  <use
    href="/static/svg/small.svg#glyph"
  />
</svg>
`;

exports[`given: Icon when: is spinning 1`] = `
.c0 {
  width: 24px;
  height: 24px;
  -webkit-animation: emHQMz 1.8s linear infinite;
  animation: emHQMz 1.8s linear infinite;
}

<svg
  className="c0"
  color=""
  rotate={null}
  size="base"
>
  <use
    href="/static/svg/base.svg#glyph"
  />
</svg>
`;
